low-liquidity-market-attack-monitor/src/agent.js

```diff
@@ -0,0 +1,165 @@
+  compTokenAddresses = compTokenAddresses.filter((addr) => excludeAddresses.indexOf(addr) === -1);
+  compTokenAddresses = compTokenAddresses.filter(
+    (addr) => !Object.keys(compTokens).includes(addr),
+  );
+
+  const promises = compTokenAddresses.map(async (tokenAddress) => {
+    const contract = new ethers.Contract(tokenAddress, compTokenAbi, provider);
+    const symbol = await contract.symbol();
+
+    // eslint-disable-next-line no-param-reassign
+    compTokens[tokenAddress] = symbol;
+  });
+  await Promise.all(promises);
+}
+
+function provideInitialize(data) {
+  return async function initialize() {
+    /* eslint-disable no-param-reassign */
+    // assign configurable fields
+    data.protocolName = config.protocolName;
+    data.protocolAbbreviation = config.protocolAbbreviation;
+    data.developerAbbreviation = config.developerAbbreviation;
+
+    const { excludeAddresses } = config;
+    data.excludeAddresses = excludeAddresses.map((addr) => addr.toLowerCase());
+
+    data.provider = getEthersBatchProvider();
+
+    const {
+      Comptroller: comptroller,
+      CompoundToken: compToken,
+    } = config.contracts;
+
+    // from the Comptroller contract, get all of the cTokens
+    const comptrollerAbi = getAbi(comptroller.abiFile);
+    data.comptrollerContract = new ethers.Contract(
+      comptroller.address,
+      comptrollerAbi,
+      data.provider,
+    );
+
+    // cToken contracts
+    data.compTokenAbi = getAbi(compToken.abiFile);
+    data.compTokens = {};
+
+    await getCompoundTokens(
+      data.provider,
+      data.comptrollerContract,
+      data.compTokenAbi,
+      data.excludeAddresses,
+      data.compTokens,
+    );
+  };
+}
+
+function provideHandleTransaction(data) {
+  return async function handleTransaction(txEvent) {
+    const {
+      protocolName,
+      protocolAbbreviation,
+      developerAbbreviation,
+      provider,
+      excludeAddresses,
+      compTokens,
+      compTokenAbi,
+      comptrollerContract,
+    } = data;
+    const findings = [];
+
+    await getCompoundTokens(
+      provider,
+      comptrollerContract,
+      compTokenAbi,
+      excludeAddresses,
+      compTokens,
+    );
+
+    const directTransferEvents = txEvent.filterLog(ERC20_TRANSFER_EVENT).filter(
+      (transferEvent) => Object.keys(compTokens).includes(transferEvent.args.to),
+    );
+
+    directTransferEvents.forEach((transferEvent) => {
+      const mintEvents = txEvent.filterLog(CERC20_MINT_EVENT, transferEvent.args.to);
+
+      mintEvents.forEach((mintEvent) => {
+        findings.push(createMarketAttackAlert(
+          protocolName,
+          protocolAbbreviation,
+          developerAbbreviation,
+          compTokens[transferEvent.args.to],
+          transferEvent.args.to,
+          mintEvent.args.mintAmount.toString(),
+          mintEvent.args.mintTokens.toString(),
+          transferEvent.args.from,
+          transferEvent.args.amount.toString(),
+        ));
+      });
+    });
```

Comment:
I don't think this is checking for the correct condition.

It looks like it is checking for the following:
- Determine if an ERC20 Transfer event occurred with any of the cToken addresses as the recipient (`to` field)
- If so, get any Mint events in the transaction that were emitted from the cToken address that was the recipient of the Transfer event

For this potential condition, I think we need to check for the following:
- Did mint() get called on any cToken contract?
  - A successful mint() call will result in a Mint event immediately followed by a Transfer event
- Then, in the same transaction, was the `underlying` token then directly transferred to that same cToken contract, which should be indicated by another Transfer event (this one won't have a Mint event right before it)?

So we need to check that a Transfer event to a cToken was emitted from the `underlying` token address (corresponding to that specific cToken).
We also need to check the ordering of the events in the log.  If the `underlying` token is transferred to the cToken contract _before_ the mint() call, that would be a botched attempt to take advantage of this issue.  That is to say, the order of the events should be:
- Mint emitted by the cToken contract
- Transfer emitted by the cToken contract
- Transfer emitted by the `underlying` token contract with the cToken address as the recipient